// fanbook-sdk@1.1.1/src/utils.ts
export const FB_WEB_SDK_EVENT = 'fb-web-sdk-event'

export function getUrlParams() {
  let queryString = window.location.search.slice(1)
  const obj: Record<string, any> = {}
  if (queryString) {
    queryString = queryString.split('#')[0]
    const arr = queryString.split('&')
    for (let i = 0; i < arr.length; i++) {
      const a = arr[i].split('=')
      const paramName = a[0]
      const paramValue = typeof a[1] === 'undefined' ? 'true' : a[1]
      // paramName = paramName.toLowerCase();
      // if (typeof paramValue === 'string') paramValue = paramValue.toLowerCase();
      if (paramName.match(/\[(\d+)?\]$/)) {
        const key = paramName.replace(/\[(\d+)?\]/, '')
        if (!obj[key]) obj[key] = []
        if (paramName.match(/\[\d+\]$/)) {
          const index = /\[(\d+)\]/.exec(paramName)?.[1]
          if (index) obj[key][index] = paramValue
        } else {
          obj[key].push(paramValue)
        }
      } else {
        if (!obj[paramName]) {
          obj[paramName] = paramValue
        } else if (obj[paramName] && typeof obj[paramName] === 'string') {
          obj[paramName] = [obj[paramName]]
          obj[paramName].push(paramValue)
        } else {
          obj[paramName].push(paramValue)
        }
      }
    }
  }
  return obj
}

export function compareVersions(v1: string, v2: string): number {
  const v1Fragments: string[] = v1.split('.')
  const v2Fragments: string[] = v2.split('.')
  let i = 0
  while (i < v1Fragments.length && i < v2Fragments.length) {
    const v1Fragment = v1Fragments[i]
    const v2Fragment = v2Fragments[i]
    if (v1Fragment != v2Fragment) {
      return (
        v1Fragment == v2Fragment ? 0
        : v1Fragment > v2Fragment ? 1
        : -1
      )
    }
    i += 1
  }
  return (
    v1Fragments.length == v2Fragments.length ? 0
    : v1Fragments.length > v2Fragments.length ? 1
    : -1
  )
}

export function getRandomString(length: 8) {
  const randomChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  let result = ''
  for (let i = 0; i < length; i++) {
    result += randomChars.charAt(Math.floor(Math.random() * randomChars.length))
  }
  return result
}

export let popupWindow: Window | null = null

export function openPopupView(url: string, target = 'fanbook-mp', width = 500, height = 800) {
  const screenWidth = window.screen.width
  const screenHeight = window.screen.height
  width = Math.min(width, screenWidth)
  height = Math.min(height, screenHeight)
  const left = (screenWidth - width) / 2
  const top = (screenHeight - height) / 2
  popupWindow = window.open(url, target, `width=${width},height=${height},top=${top},left=${left}`) //,noopener=yes
  return popupWindow
}

export function openPopupIframe(url: string, width = 500, height = 800) {
  // 检查是否已存在相同的 iframe
  let target = document.getElementById('popupWindow') as HTMLIFrameElement
  if (target && target instanceof HTMLIFrameElement) {
    target.src = url
    return target
  }

  // 创建 iframe 元素
  target = document.createElement('iframe')

  // 设置 iframe 的属性
  target.id = 'popupWindow'
  target.src = url
  const screenWidth = window.innerWidth
  const screenHeight = window.innerHeight
  width = Math.min(width, screenWidth)
  height = Math.min(height, screenHeight)
  const left = (screenWidth - width) / 2
  const top = (screenHeight - height) / 2
  const styles = {
    position: 'fixed',
    width: `${width}px`,
    height: `${height}px`,
    left: `${left}px`,
    top: `${top}px`,
    backgroundColor: '#fff',
    border: '1px solid #ccc',
    borderRadius: '8px',
    zIndex: 9999,
  }
  Object.assign(target.style, styles)
  document.body.appendChild(target)
  return target
}
